### PR TITLE
Show preview of WordPress embed in editor

### LINKED ISSUE #21029: Bug: Link sometimes embeds other times not.
WordPress 5.3.2
Using Gutenberg version 7.7.1. (same affect when deactivating the Gutenberg plugin.)
Twenty Twenty theme.

**Describe the bug**
Adding a link through pasting a link into a paragraph block sometimes makes it into an embedded link and other times a text link.

**To reproduce**
1. In a paragraph block paste this link: https://make.wordpress.org/
2. Result should be an embedded link.

Screen Shot 2020-03-19 at 21 40 04

-------

1. In a paragraph block paste this link: https://make.wordpress.org/core/
2. An Embed URL dialog box shows up saying "Sorry, this content could not be embedded."

Screen Shot 2020-03-19 at 21 41 06

I click the Convert to (text) link.
Screen Shot 2020-03-19 at 21 42 52

The result on the frontend showing an embed and a text link:
Screen Shot 2020-03-19 at 21 43 54

Here is an example where I also pasted the following link into a paragraph block:  https://make.wordpress.org/test/2020/03/11/gutenberg-usability-testing-for-january-2020/
Frontend result showing three different visuals.

Screen Shot 2020-03-19 at 21 53 32

**Expected behavior**
I expect to be able to choose when or if I want a text link to be embedded. 

If a link wants to embed, then I expect a dialog screen to show up similar to this mockup.
Embed-link-option-Gutenberg

**Desktop (please complete the following information):**
 - OSX 10.13.6
 - Browser Chrome
 - Gutenberg version 7.7.1

Plugin support issues:
https://wordpress.org/support/topic/paste-links-in-gutenberg-without-embedding/
https://wordpress.org/support/topic/paste-link-without-creating-an-embed-block/

### LINKED ISSUE #21945: WordPress embed block is not hiding the blockquote
**Describe the bug**
WordPress oembeds provide a blockquote fallback for cases where the iframe cannot be rendered ( see https://build.trac.wordpress.org/browser/trunk/wp-includes/embed.php?rev=47325#L449 and https://build.trac.wordpress.org/browser/trunk/wp-includes/js/wp-embed.js?rev=43426#L50 ).

In classic editor, this feature works as expected, but in Gutenberg / Block editor the blockquote can still be seen.

**To reproduce**
1. Go to a new post screen in wp-admin
2. Enter an embeddable link of your own site 
3. Note the post title being above the iframe

**Expected behavior**
The blockquote should be hidden as it is in the classic editor.

**Screenshots**
classic-editor-with-hidden-blockquote
gutenberg-with-visible-blockquote

**Editor version (please complete the following information):**
- WordPress version: 5.5-alpha-47572
- Does the website has Gutenberg plugin installed, or is it using the block editor that comes by default? Reprodubible in both, default and gutenberg plugin
- If the Gutenberg plugin is installed, which version is it? 4.8.1

**Desktop (please complete the following information):**
 - OS: macOS
 - Browser: chrome
 - Version: 81.0.4044.122

### LINKED ISSUE #25344: Embeds: WordPress Embeds not properly previewed in Editor
**Describe the bug**
WordPress Embeds are rendered quite differently in the Editor than on the frontend.

**To reproduce**
1. Start a new post.
2. Insert the URL of a WordPress site, e.g. https://ma.tt/2020/07/corner-office-interview/
3. Verify that rather than the embed, only a link with the page title is inserted.

**Expected behavior**
The WordPress Embed should be inserted, as seen on the frontend.

**Screenshots**
Editor:
image

Frontend:
image

**Editor version (please complete the following information):**
- WordPress version: 5.5
- Does the website has Gutenberg plugin installed, or is it using the block editor that comes by default? gutenberg plugin
- If the Gutenberg plugin is installed, which version is it? 8.9

**Additional context**
Found by @david-szabo97 here: https://github.com/Automattic/wp-calypso/issues/39935#issuecomment-686524172. Quoting that comment:

> EDIT:
> ~**TLDR: We need this patch: D48267-code**~
> **TLDR: oAuth embed auto discovery embeds the iframe**
> 
> Figured it out. It's actually being embedded:
> image
> But you can see an inline style there on the iframe: `position: absolute;clip: rect(1px, 1px, 1px, 1px);` which is basically hiding the embedded content.
> image
> 
> `wp-embed.js` is supposed to remove that inline style, but since we don't have that on WP.com, it's not happening.

So it seems we need to load that `wp-embed.js` in the editor to make things work? :thinking: 

This issue also discussed as part of https://github.com/WordPress/gutenberg/issues/21029, and in https://github.com/bobbingwide/oik/issues/158 (although with a somewhat different conclusion, IIRC).

cc/ @mcsf @ntsekouras @paaljoachim @bobbingwide

### PR DESCRIPTION
**Pull request by ntsekouras**
## Description
Fixes: https://github.com/WordPress/gutenberg/issues/25344
Fixes: https://github.com/WordPress/gutenberg/issues/21945
Fixes: https://github.com/WordPress/gutenberg/issues/21029

Great description in this issue about the root of the problem in this issue: https://github.com/WordPress/gutenberg/issues/25344

## How has this been tested?
Locally

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Review by mcsf — COMMENTED** [review]
I like this!

Can you conceive a test for this?

**Comment by ntsekouras** [comment]
It's not complete. The problem here is that the responsiveness of the `iframe` at front-ent, is handled from themes. Tested with many default themes like 2020 etc..

In the editor is not and now it has bigger height, leaving a gap between the iframe and the caption. Aside the incorrect view, if you click the gap and throws an error.

I have made some experiments, one of which and IMO the best one, is to attach a `load` event to the rendered iframe and set proper height. I haven't been successful yet, but I'll try it a bit more tomorrow.

Any ideas for this are welcome. :)

**Comment by ntsekouras** [comment]
>Can you conceive a test for this?

Actually all embedding tests are mocked (their response). One reason for this should be not to have to handle the case this PR handles, to attach listeners etc for iframe content loading..

File: packages/block-library/src/embed/wp-embed-preview.js [hunk 1]
```diff
@@ -18,6 +20,36 @@ class WpEmbedPreview extends Component {
...
+	 */
+	resizeWPembeds( { data: { secret, message, value } = {} } ) {
+		if ( [ secret, message, value ].some( ( attribute ) => ! attribute ) )
+			return;
+
+		if ( message === 'height' ) {
+			const iframes = document.querySelectorAll(
+				`iframe[data-secret="${ secret }"`
+			);
+			( iframes || [] ).forEach( ( iframe ) => {
```
Comment (mcsf) [inline packages/block-library/src/embed/wp-embed-preview.js]:
```suggestion
			iframes.forEach( ( iframe ) => {
```

Doesn't seem like `querySelectorAll` ever returns anything other than a NodeList.

File: packages/block-library/src/embed/wp-embed-preview.js [hunk 2]
```diff
@@ -38,15 +70,21 @@ class WpEmbedPreview extends Component {
 
 	render() {
 		const { html } = this.props;
-		const doc = new DOMParser().parseFromString( html, 'text/html' );
-		doc.querySelector( 'iframe' ).removeAttribute( 'style' );
-		doc.querySelector( 'blockquote' ).style.display = 'none';
+		if ( this.node.current ) {
```
Comment (mcsf) [inline packages/block-library/src/embed/wp-embed-preview.js]: In terms of life cycle, is this the right place to do this?

What about using callback refs like:

```js
function processContents( ref ) {
  if ( ref ) {
    ref.querySelector( ... ).style.display = 'none';
  ...
}

<div ref={ processContents }>...</div>

File: packages/block-library/src/embed/wp-embed-preview.js [hunk 3]
```diff
@@ -18,6 +20,36 @@ class WpEmbedPreview extends Component {
...
+		window.removeEventListener( 'message', this.resizeWPembeds );
+	}
+
+	/**
+	 * Checks for WordPress embed event signaling the height
+	 * change when content loads.
+	 *
+	 * @param {WPSyntheticEvent} event Message event.
+	 */
+	resizeWPembeds( { data: { secret, message, value } = {} } ) {
```
Comment (mcsf) [inline packages/block-library/src/embed/wp-embed-preview.js:37]: This all reads a bit like dark magic. :) I had never come across this height signalling before.

I'm not too confident to approve the change.

In any case, providing some context and linking to adequate prior art would be nice.

**Comment by paaljoachim** [comment]
I test this locally and it is showing a nice progress!
There are just a few things I notice at the moment.

The caption toolbar is seen as part of the embed preview as if it is an image block.
Clicking the edit icon and then clicking the embed button will for a while show the old embed preview and suddenly it will show the new.

File: packages/block-library/src/embed/wp-embed-preview.js (hunk 3 above)
Comment (ntsekouras) [reply packages/block-library/src/embed/wp-embed-preview.js:43]: You're right! I've updated the jsdoc.

**Comment by ntsekouras** [comment]
@paaljoachim 
>The caption toolbar is seen as part of the embed preview as if it is an image block.
Can you elaborate on that?

>Clicking the edit icon and then clicking the embed button will for a while show the old embed preview and suddenly it will show the new.

You're right here, it needs fixing. I'll look into it

**Comment by paaljoachim** [comment]
Hey @ntsekouras 
Adding the embed this is the result that I see:
Screen Shot 2020-09-21 at 11 20 26

As if the embed was an image block.

**Comment by ntsekouras** [comment]
@paaljoachim I fixed the edit problem now.

Regarding your screenshot, I don't see anything weird there..
Can you create a `youtube` embed for example? Do you see something different in toolbars etc?

**Comment by paaljoachim** [comment]
It sounds like you are saying that the caption should be there for a WP embed. 
I thought it was a bug. That the caption below the WP embed was not supposed to be there.

**Review by mcsf — APPROVED** [review]
Let's merge this. :)
